design-documents/tracking-tool.md

```diff
@@ -0,0 +1,80 @@
+# Tech Upgrade Strategy
+
+## Overview
+
+Magento uses different 3rd party components and libraries like Symfony, Zend, hash generators, SDK, etc. And during Magento release we need to update components versions manually. Also, updates to the latest package version might require some code migration in Magento, which might impact release dates.
+To avoid such problem, we need to have a mechanism to track latest versions of 3rd party components.
+
+As most of dependencies specified in `composer.json`, such process can be automated.
+
+## Solution
+
+Dependencies, specified in `composer.json`, can be parsed and according to the dependencies list, we can find the latest version of the package.
+At least, we should have a tool which can find the latest version based on provided list and criteria.
+
+### Requirements
+
+ - An ability to provide the list of components for update
+ - An ability to recognized dependencies in `composer.json`
+ - Find the latest version according to the specific rule in multiple sources
+ - A possibility to extend the packages source list (Packagist, Github, Gitlab)
+ - Specific update rules like `^`, `~`, `-`, `*`, etc.
+ - A notification mechanism for available updates
+ 
+### Packages source 
+
+The most of packages are specified in `composer.json` but some of them are not specified, like database, or can be added in the future not via composer.
+So we need to provide a possibility to extend source of packages versions.
+
+The tool should support multiple input types like CLI, configuration files, URL to composer.json file.
+
+#### Github REST API
+
+Unfortunately, [Packagist API](https://packagist.org/apidoc) does not provide a possibility to retrieve the package latest version but, as the most of packages are hosted on Github, [Github REST API](https://developer.github.com/v3/repos/releases/) might be used for such purposes.
```

Comment:
I see on your link to Packagist API that it allows to get package metadata which contains all available package versions: https://packagist.org/apidoc#get-package-data
Seems like it’s what you’re looking for as it could be parsed and compared against defined update rules like `^`

Reply:
And they have field called “version_normalized” which, as I understand, allows you to not parse all those prefixes on your own

Reply:
Not exactly, because you should have a possibility to find different versions like the latest patch, minor, major.

Reply:
> And they have field called “version_normalized” which, as I understand, allows you to not parse all those prefixes on your own

I have seen it and probably it can be used but I didn't find the information why it contains 4 digits.

Reply:
Yes, I just wanted to point out that this line is not completely correct. And in my opinion it’s better to stay close to the way how composer gets its info as it’s how we’d finally get the dependency. It could be published on GitHub, but not available on composer for example - not sure ho composer would be looking for dependency in such case.
Also, probably we could avoid implementing and supporting some things if we’d be close to composer - it already has everything we want, probably not always behind the right interface.
Here’s an example of the composer plugin I’m using which doing such checks: https://github.com/soullivaneuh/composer-versions-check

Reply:
> A package could be published on GitHub, but not available on composer for example - not sure ho composer would be looking for dependency in such case.

Mysql it's a good example, we don't have dependency in `composer.json` but using it. This proposal about all dependencies which exist not only in `composer.json`.

> composer-versions-check is a plugin for Composer.
It warns user for outdated packages from last major versions after update command.

We are interested not only in major versions. Not always we can use the latest version, as an example some zend components but anyway the have to use the latest compatible version.